internal/rootcoord/root_coord.go

```diff
@@ -1048,6 +1051,201 @@ func (c *Core) DropCollection(ctx context.Context, in *milvuspb.DropCollectionRe
 	return merr.Success(), nil
 }
 
+// TruncateCollection truncate collection
+func (c *Core) TruncateCollection(ctx context.Context, in *rootcoordpb.TruncateCollectionRequest) (*commonpb.Status, error) {
+	if err := merr.CheckHealthy(c.GetStateCode()); err != nil {
+		return merr.Status(err), nil
+	}
+
+	metrics.RootCoordDDLReqCounter.WithLabelValues("TruncateCollection", metrics.TotalLabel).Inc()
+	tr := timerecord.NewTimeRecorder("TruncateCollection")
+	log := log.Ctx(ctx).With(zap.String("role", typeutil.RootCoordRole),
+		zap.String("dbName", in.GetDbName()), zap.String("collectionName", in.GetCollectionName()),
+		zap.Bool("needLoad", in.NeedLoad), zap.Bool("clearTempCollection", in.Clear))
+	log.Info("received request to truncate collection")
+
+	if in.Clear {
+		log.Debug("try to remove the temp collection")
+		return c.dropCollectionForTempSync(ctx, in)
+	}
+
+	log.Debug("create a temp collection with the same schema as target collection")
+	ct := &copyCollectionTask{
+		baseTask:       newBaseTask(ctx, c),
+		Req:            in,
+		collectionName: util.GenerateTempCollectionName(in.CollectionName),
+	}
+	if err := c.scheduler.AddTask(ct); err != nil {
+		log.Info("failed to enqueue request to create a temp collection", zap.Error(err), zap.Uint64("ts", ct.GetTs()))
+		metrics.RootCoordDDLReqCounter.WithLabelValues("TruncateCollection", metrics.FailLabel).Inc()
+		return merr.Status(err), nil
+	}
+
+	if err := ct.WaitToFinish(); err != nil {
+		log.Info("failed to create a temp collection", zap.Error(err), zap.Uint64("ts", ct.GetTs()))
+		metrics.RootCoordDDLReqCounter.WithLabelValues("TruncateCollection", metrics.FailLabel).Inc()
+		return merr.Status(err), nil
+	}
+
+	t := &truncateCollectionTask{
+		baseTask: newBaseTask(ctx, c),
+		Req:      in,
+	}
+	defer func() {
+		log.Debug("drop the temp collection async")
+		c.dropCollectionForTemp(ctx, in, t.collectionID, t.tempCollectionID)
```

Comment:
Why we should do this even if the truncating operation is done?

Reply:
I think we can provide a abstraction which combine the undoTask and redoTask for the truncating workflow, like below:

```go
type generalTask struct{
  undoTask
  redoTask
}

func (t *generalTask) Execute() error {
  ReturnErrIf(t.undoTask.Execute())
  go t.redoTask.Execute()
  return nil
}
```

And truncating a collection in fact can be simplified to:
```go
undo.AddStep(CreateTempCollection, RemoveTempCollection) // 1
undo.AddStep(CreateIndexesForCollection, nullstep) // 2
undo.AddStep(LoadTempCollection, nullstep) // 3
undo.AddStep(ExchangeCollections, nullstep) // 4
redo.AddStep(RemoveOriginalCollection) // 5
```

Consider the recovery and atomicity:

- Fail at step 1, no temp collection was created;
- Fail at step 2, the step executor will execute the undo step of 1, RemoveTempCollection, which will drop the indexes and releasing the temp collection, rootcoord will also do this when it starts;
- Fail at 3 or 4, same with 2;
- We can return when 4 is succeeded, because 4 is atomic so no any side effect will be brought;
- After 4 is done, the gc of original collection is asynchronous;

Reply:
Hi @longjiquan , I'm working on this feature now. Would like to follow up your comment here by a few questions that popped up to me, whilst implementing `truncateCollection()` at the `root_coord` node


> undo.AddStep(CreateTempCollection, RemoveTempCollection)

I noticed both of those `steps` will be very similar to `create_collection_task` (and `drop_collection_task` respectively). So when I implement this, I can't decide between 2 options:
  1. Create a new `create(drop)CollectionStep`, and copy almost of the logic from `create(drop)_collection_task` 
  2. Schedule a `create_collection_task` for creating temp collection, before scheduling `truncate_collection_task`, if the `truncate_collection_task` fail, will schedule a `drop_collection_task` for created temp collection
  
  Which option would you prefer more ? Personally, I lean toward the 2nd, because it will make the codebase less repetitive.  
  
  
  > redo.AddStep(RemoveOriginalCollection) // 5
  
  I do have newbie question on this. Actually we don't have to add this step right ? Because in the step 4, we will exchange the collection, then mark the original collection to `drop` state. Then the async background GC will eventually remove the original ones. Am I right ?